src/bms/parse/validity.rs

```diff
@@ -0,0 +1,614 @@
+//!
+//! This module provides a set of structural validations that are independent of
+//! the parsing process. It can be used after editing `Bms` in-memory to ensure
+//! referential integrity and basic invariants required for correct playback.
+
+use thiserror::Error;
+
+use crate::bms::command::{
+    ObjId,
+    channel::{Key, NoteKind, PlayerSide},
+    time::ObjTime,
+};
+
+use super::model::{Bms, obj::Obj};
+
+/// Empty-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityEmpty {
+    /// Text content is empty at the time.
+    #[error("Empty text content at {0:?}")]
+    EmptyTextAt(ObjTime),
+    /// Option content is empty at the time.
+    #[cfg(feature = "minor-command")]
+    #[error("Empty option content at {0:?}")]
+    EmptyOptionAt(ObjTime),
+}
+
+/// Missing-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityMissing {
+    /// A note references an ObjId without a corresponding WAV definition.
+    #[error("Missing WAV definition for note object id: {0:?}")]
+    MissingWavForNote(ObjId),
+    /// A BGM references an ObjId without a corresponding WAV definition.
+    #[error("Missing WAV definition for BGM object id: {0:?}")]
+    MissingWavForBgm(ObjId),
+    /// A BGA change references an ObjId without a corresponding BMP/EXBMP definition.
+    #[error("Missing BMP definition for BGA object id: {0:?}")]
+    MissingBmpForBga(ObjId),
+}
+
+/// Mapping-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityMapping {
+    /// The per-key time index (`ids_by_key`) is missing or mismatched for an existing note.
+    #[error(
+        "Key index mismatch for note at time {time:?} (side={side:?}, key={key:?}), expected mapping to {expected:?}"
+    )]
+    IdsByKeyMismatch {
+        /// Player side where the note is placed.
+        side: PlayerSide,
+        /// Key lane where the note is placed.
+        key: Key,
+        /// Timestamp of the note.
+        time: ObjTime,
+        /// Expected object id that should be mapped at this timestamp.
+        expected: ObjId,
+    },
+    /// Orphan entry found in `ids_by_key`: mapping exists but no actual note at that time.
+    #[error(
+        "Orphan key index mapping at time {time:?} (side={side:?}, key={key:?}), mapped to {mapped:?} but note not found"
+    )]
+    IdsByKeyOrphanMapping {
+        /// Player side where the mapping is placed.
+        side: PlayerSide,
+        /// Key lane where the mapping is placed.
+        key: Key,
+        /// Timestamp of the mapping.
+        time: ObjTime,
+        /// Mapped object id.
+        mapped: ObjId,
+    },
+}
```

Comment:
This check comes from lack of encapsulation of `Bms`, so it should be removed. We need to improve `Bms`'s APIs not to happen such situations.